File: dsa/dfs/841_Keys_and_Rooms.java
package dfs;
import java.util.*;

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        Stack<Integer> stack = new Stack<Integer>();
              
        for (int i = 0; i < rooms.get(0).size(); i++){
            stack.push(rooms.get(0).get(i));
            visited[0] = true;
        }
        int countVisited = 1;

        while (!stack.isEmpty()){
            int curNode = stack.pop();
            if (!visited[curNode]){
                visited[curNode] = true;
                countVisited += 1;
                for (int i = 0; i < rooms.get(curNode).size(); i++){
                    stack.push(rooms.get(curNode).get(i));
                }
            }
        }

        if (countVisited == rooms.size()){
            return true;
        }
        else {
            return false;
        }
    }
}

File: dsa/13.dfs/841_Keys_and_Rooms.py
import typing as t

class Solution:
    def canVisitAllRooms(self, rooms: t.List[t.List[int]]) -> bool:
        stack: t.List[int] = [0]
        visited: t.Set[int] = set()
        while stack:
            cur: int = stack.pop()
            if cur in visited:
                continue
            
            visited.add(cur)
            for adj in rooms[cur]:
                if adj not in visited:
                    stack.append(adj)
        
        if len(visited) == len(rooms):
            return True
        else:
            return False